File: Pruebas-Unitarias/Ejemplo-Python/src/appointment_service.py
class AppointmentService:
    def __init__(self, appointment_repo, clock):
        # Inyecta reloj y repo para controlar tiempo y persistencia en tests.
        self.appointment_repo = appointment_repo
        self.clock = clock

    def schedule_appointment(self, patient_id, starts_at):
        # Regla 1: la cita debe ser en el futuro.
        now = self.clock.now()
        if starts_at <= now:
            raise ValueError("APPOINTMENT_MUST_BE_FUTURE")

        # Regla 2: no debe existir conflicto para el mismo paciente y hora.
        has_conflict = self.appointment_repo.has_conflict(patient_id, starts_at)
        if has_conflict:
            raise ValueError("APPOINTMENT_CONFLICT")

        # Guardar si pasa las validaciones.
        return self.appointment_repo.save({"patient_id": patient_id, "starts_at": starts_at})

File: Pruebas-Unitarias/Ejemplo_Node/src/appointmentService.js
function buildAppointmentService({ appointmentRepo, clock }) {
  return {
    // Caso de uso: programar cita para un paciente.
    // Regla 1: la cita debe ser en el futuro.
    // Regla 2: no debe existir conflicto de horario para ese paciente.
    async scheduleAppointment({ patientId, startsAt }) {
      // Se usa clock inyectado para hacer pruebas deterministicas.
      const now = clock.now();
      if (startsAt <= now) {
        throw new Error("APPOINTMENT_MUST_BE_FUTURE");
      }

      // Consulta de conflicto (normalmente en DB, aqui depende del repo inyectado).
      const hasConflict = await appointmentRepo.hasConflict(patientId, startsAt);
      if (hasConflict) {
        throw new Error("APPOINTMENT_CONFLICT");
      }

      // Si pasa validaciones, se guarda la cita.
      return appointmentRepo.save({ patientId, startsAt });
    }
  };
}

module.exports = { buildAppointmentService };
